client.js

```diff
@@ -53,21 +51,36 @@ module.exports = function (opts, cb) {
   ])
 
   ms.client(remote, function (err, stream) {
-    if(err) return cb(explain(err, 'could not connect to sbot'))
-    var id = '@'+stream.remote.toString('base64')+'.ed25519'
-    // mix: id === config.keys.id, note sure why this is needed
-    var sbot = muxrpc(manifest, false, null, id)
+    if(err) {
+      return cb(explain(err, 'could not connect to sbot'))
+    }
 
-    // fix blobs.add. (see ./util/fix-add-blob.js)
-    if (sbot.blobs && sbot.blobs.add)
-      sbot.blobs.add = fixBlobsAdd(sbot.blobs.add)
+    // HACK: This refers to the *remote* feed ID, not our local ID, and we don't
+    // actually have a way of determining the feed type. This is fine when all
+    // feeds are ed25519, but will cause problems in the future.
+    const remoteId = `@${stream.remote.toString('base64')}.ed25519`
 
-    pull(stream, sbot.stream, stream)
+    const sbot = muxrpc((err) => {
+      if (err) {
+        return cb(err)
+      }
+
+      // fix blobs.add. (see ./util/fix-add-blob.js)
+      if (sbot.blobs && sbot.blobs.add) {
+        sbot.blobs.add = fixBlobsAdd(sbot.blobs.add)
+      }
 
-    // mix: I've added this line because somewhere in the stack the id property was removed
-    // and it's broken a bunch of code I've written
-    if (!sbot.id) sbot.id = id
+      // This refers to the *local* feed ID, and gives us a quick and easy way
+      // to reference ourselves without calling `whoami()`. These may share a
+      // value when the client and server share an identity, but won't be the
+      // same when connecting to a remote server.
+      if (sbot.id == null) {
+        sbot.id = opts.keys.id
+      }
```

Comment:
Would also love for someone to double-check this change and its documentation to make sure they're right.

Reply:
(This was previously `sbot.id = remoteId`, which I think was wrong (?).)

Reply:
why do you think that is wrong? it's useful to know who you are talking too... I feel there is probably some code somewhere that uses that.

Reply:
The code was added in https://github.com/ssbc/ssb-client/pull/50/commits/7906cb364ceff0c558d8d69c74146888ff8cd6fb and I think either the comments are backward, the logic is backward, or I'm misreading something.

That commit defines `id` as the remote ID, comments `id === config.keys.id`, and puts the remote in `sbot.id`, which I thought was supposed to be the same as `sbot.whoami().id` (the local ID). Could you help be me less confused? I wouldn't be surprised if I'm just reading it wrong.

Reply:
there is the server's id, the clients id. whoami returns the servers id. on a local instance of sbot, sbot.id is the server id... so `rpc_sbot.id` ought to be the remote id? that would make the it possible for the same code to work on a local or remote instance of sbot. also note, that when a rpc function is called it's called with the context `{id: remote_id}` so that `this.id == remote_id` which makes it possible to know which peer called this function. That is used for implementing blocks for example. That's my understanding of it. It's up otherwise up to clients to keep track of their own id.

Reply:
Thanks for the info. I've just pushed a commit to fix this and make sure it's set as the remote ID. Please let me know if you'd like to see any other changes!